raft/src/main/java/org/apache/kafka/raft/QuorumState.java

```diff
@@ -402,58 +417,100 @@ public void transitionToUnattached(int epoch) {
     }
 
     /**
-     * Grant a vote to a candidate. We will transition/remain in Unattached
-     * state until either the election timeout expires or a leader is elected. In particular,
-     * we do not begin fetching until the election has concluded and
-     * {@link #transitionToFollower(int, int, Endpoints)} is invoked.
+     * Grant a vote to a candidate as Unattached. We will transition to Unattached with votedKey
+     * state and remain there until either the election timeout expires or we discover the leader.
      */
-    public void transitionToUnattachedVotedState(
+    public void unattachedAddVotedState(
         int epoch,
         ReplicaKey candidateKey
     ) {
         int currentEpoch = state.epoch();
         if (localId.isPresent() && candidateKey.id() == localId.getAsInt()) {
             throw new IllegalStateException(
                 String.format(
-                    "Cannot transition to Voted for %s and epoch %d since it matches the local " +
+                    "Cannot add voted key (%s) to current state (%s) in epoch %d since it matches the local " +
                     "broker.id",
                     candidateKey,
+                    state,
                     epoch
                 )
             );
         } else if (localId.isEmpty()) {
-            throw new IllegalStateException("Cannot transition to voted without a replica id");
-        } else if (epoch < currentEpoch) {
+            throw new IllegalStateException("Cannot add voted state without a replica id");
+        } else if (epoch != currentEpoch || !isUnattachedNotVoted()) {
```

Comment:
this method is solely for the case of adding voted state to unattached in the same epoch (if higher epoch, we take the path of transitionToUnattached)

the transitionToUnattached method has the following comment, I will duplicate it for this method too
```
     * Note, if we are transitioning from unattached and there is no epoch change, we take the path of
     * unattachedAddVotedState instead.
```

Reply:
I could also remove `epoch` as one of the params (method can use state.currentEpoch instead of parameter value). But I would prefer to keep `epoch` as a param so we can validate the method is being used correctly (without any unintentional epoch change).

Reply:
discussed offline that this method is only used for adding voted state to unattached (in same epoch)

Reply:
Sounds good to keep the epoch parameter and validating it against the current epoch.